### PR TITLE
fix(core): fix symbol capacity displayed wrong after restart in show columns queries

### PR DESCRIPTION
**Pull request by ideoma**
If a SYMBOL column is added/changed the type and then QuestDB restarted then `SHOW COLUMNS` query returns random values as the symbol capacity. This is due to the fact that symbol capacity was not stored in `_meta` file or at least not on every DDL operation.

This PR changes to store the symbol capacity in the `_meta` file along with the symbol cached flag. It also introduces an additional Integer field at offset `META_OFFSET_META_FORMAT_MINOR_VERSION` that indicates if these values were properly stored so that `MetadataCache` can rely on them.

`MetadataCache` is changed to read symbol capacities directly from the symbol files when `META_OFFSET_META_FORMAT_MINOR_VERSION` does not match the expected latest value.

Also, `TableWriter` is heavily refactored along the paths of DDL changes, so there is only one method to serialize metadata from the memory object model into a file. This reduces the code maintenance problems like not updating every places where column metadata is re-written that happened with the symbol capacity and has happened few times before.

Also fixes index files not dropped as reported in #5259

### CODE REVIEW

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 307 / 339 (90.56%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/MetadataCache.java|42|50|84.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|211|233|90.56%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/VacuumColumnVersions.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/tasks/ColumnPurgeTask.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|

File: core/src/main/java/io/questdb/cairo/TableWriter.java [hunk 1]
```diff
@@ -946,11 +933,12 @@ public void changeColumnType(
...
     ) {
 
-        int existingColIndex = metadata.getColumnIndexQuiet(columnName);
+        int existingColIndex = metadata.getColumnIndexQuiet(name);
         if (existingColIndex < 0) {
             throw CairoException.nonCritical().put("cannot change column type, column does not exists [table=")
-                    .put(tableToken.getTableName()).put(", column=").put(columnName).put(']');
+                    .put(tableToken.getTableName()).put(", column=").put(name).put(']');
         }
+        String columnName = metadata.getColumnName(existingColIndex);
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TableWriter.java:941]: Yes, incoming `CharSequence` can be re-used while the column name is passed to purge column job, hence it's better to use a stable `String` object from this point on.

File: core/src/main/java/io/questdb/cairo/MetadataCache.java [hunk 2]
```diff
@@ -270,6 +288,55 @@ private void hydrateTable0(@NotNull TableToken token, boolean throwError) {
...
+
+            // get column version
+            Path tablePath = path.trimTo(configuration.getRoot().length()).concat(token);
+            int rootLen = tablePath.size();
+            tablePath.concat(TableUtils.COLUMN_VERSION_FILE_NAME);
+
+            final long columnNameTxn;
+            FilesFacade ff = configuration.getFilesFacade();
+            try (columnVersionReader) {
+                columnVersionReader.ofRO(ff, path.$());
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/MetadataCache.java:310]: good one, I added a close

File: core/src/main/java/io/questdb/cairo/MetadataCache.java [hunk 3]
```diff
@@ -270,6 +288,55 @@ private void hydrateTable0(@NotNull TableToken token, boolean throwError) {
...
+
+                columnVersionReader.readUnsafe();
+                columnNameTxn = columnVersionReader.getDefaultColumnNameTxn(writerIndex);
+            }
+
+            // use txn to find correct symbol entry
+            final var offsetFileName = TableUtils.offsetFileName(tablePath.trimTo(rootLen), columnName, columnNameTxn);
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/MetadataCache.java]: I removed `tablePath` alias

File: core/src/main/java/io/questdb/cairo/TableUtils.java [hunk 4]
```diff
@@ -221,6 +223,10 @@ public static void allocateDiskSpaceToPage(FilesFacade ff, long fd, long size) {
         allocateDiskSpace(ff, fd, size);
     }
 
+    public static int calculateMetadataMinorFormatVersion(int metadataVersion) {
+        return Numbers.encodeLowHighShorts(Numbers.decodeLowShort(metadataVersion), META_MINOR_VERSION_LATEST);
+    }
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TableUtils.java]: done

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 309 / 341 (90.62%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/MetadataCache.java|43|51|84.31%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|212|234|90.60%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/VacuumColumnVersions.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/tasks/ColumnPurgeTask.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 309 / 341 (90.62%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/MetadataCache.java|43|51|84.31%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|212|234|90.60%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/VacuumColumnVersions.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/tasks/ColumnPurgeTask.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 309 / 341 (90.62%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/MetadataCache.java|43|51|84.31%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|212|234|90.60%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/VacuumColumnVersions.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/tasks/ColumnPurgeTask.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 304 / 337 (90.21%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/MetadataCache.java|43|51|84.31%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|214|237|90.30%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/VacuumColumnVersions.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/ColumnPurgeTask.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 303 / 339 (89.38%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/MetadataCache.java|43|51|84.31%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|213|239|89.12%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/VacuumColumnVersions.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/ColumnPurgeTask.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 303 / 339 (89.38%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/MetadataCache.java|43|51|84.31%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|213|239|89.12%|
|:large_blue_circle:|io/questdb/cairo/TableConverter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/VacuumColumnVersions.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/PurgingOperator.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetadata.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/ConvertOperatorImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/ColumnPurgeTask.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/DropIndexOperator.java|1|1|100.00%|

**Pull request merged**